source/material_model/steinberger.cc

```diff
@@ -132,9 +132,53 @@ namespace aspect
     void
     Steinberger<dim>::initialize()
     {
+      // This model allows the user to provide several PerpleX P-T lookup files,
+      // each of which corresponds to a different material.
+      // The thermodynamic properties and stable mineral phases within each material
+      // will in general be different, and must be averaged in a reasonable way.
+
+      // In the following, we populate the unique_phase_names vector.
+      // We do this because we are choosing in this material model to combine
+      // minerals with different compositions but which are the same phase.
+      std::set<std::string> set_phase_volume_column_names;
       for (unsigned i = 0; i < material_file_names.size(); i++)
-        material_lookup.push_back(std_cxx14::make_unique<MaterialModel::MaterialUtilities::Lookup::PerplexReader>
-                                  (data_directory+material_file_names[i],interpolation,this->get_mpi_communicator()));
+        {
+          material_lookup.push_back(std_cxx14::make_unique<MaterialModel::MaterialUtilities::Lookup::PerplexReader>
+                                    (data_directory+material_file_names[i],interpolation,this->get_mpi_communicator()));
+
+          // Here we look up all of the column names and insert
+          // unique names into the unique_phase_names vector.
+          std::vector<std::string> phase_volume_column_names = material_lookup[i]->phase_volume_column_names();
+          for (std::string const &name : phase_volume_column_names)
+            {
+              if (std::find(unique_phase_names.begin(),
+                            unique_phase_names.end(),
+                            name) == unique_phase_names.end())
+                unique_phase_names.push_back(name);
+            }
+        }
+
+      // We now fill the unique_phase_indices object
+      // with the unique_phase_names index for each of the phases
+      // stored in each material lookup.
+
+      // Initialise the column indices
+      unique_phase_indices.resize(material_file_names.size(),
+                                  std::vector<unsigned int>());
+
+      for (unsigned int i = 0; i < material_file_names.size(); i++)
+        {
+          std::vector<std::string> phase_volume_column_names = material_lookup[i]->phase_volume_column_names();
+          for (unsigned int j = 0; j < phase_volume_column_names.size(); j++)
```

Comment:
Same here: you now loop over all j twice, but you can combine this with the loop over the strings above. (At least I am thinking this will make the code shorter, if that just makes it more complicated, it's not worth changing).

Reply:
Same problem as above. The two loops need to be separated, because the second loop requires a completed `unique_phase_names` object from the first.

Reply:
Not a problem!